draft-bishop-httpbis-http2-additional-certs.md

```diff
@@ -428,18 +430,32 @@ Client                                      Server
    -- (stream 0) CERTIFICATE ------------------>
    -- (stream N) USE_CERTIFICATE -------------->
    <----------------------- (stream N) 200 OK --
-
 ~~~
 {: #ex-http2-client-requested title="Reactive Certificate Authentication"}
 
-A server SHOULD provide certificates for an origin before pushing resources from
-it or supplying content referencing the origin. If a client receives a
-`PUSH_PROMISE` referencing an origin for which it has not yet received the
-server's certificate, the client MUST verify the server's possession of an
-appropriate certificate by sending a `CERTIFICATE_NEEDED` frame on the pushed
-stream to inform the server that progress is blocked until the request is
-satisfied. The client MUST NOT use the pushed resource until an appropriate
-certificate has been received and validated.
+On subsequent requests, the client can proactively indicate that an existing
+certificate MAY be used by including a `USE_CERTIFICATE` frame as part of the
+request.  If the indicated certificate does not correspond to the request the
```

Comment:
I think that we should require the client to send the USE_CERTIFICATE frame before the DATA frames that convey the request body.

This is because we need a way to cap the amount of the memory being required to buffer the request. In CGI and other gateway protocols, the properties of the client certificate needs to sent before the POST body.

By requiring the client to send USE_CERTIFICATE before DATA frames, the upper limit of buffer size can be capped to the size of the stream-level flow control window.

I also think that it would also be a good idea to require the client to _always_ send the USE_CERTIFICATE frame within the request. A server will be permitted to send a CERTIFICATE_NEEDED frame if the client sends an empty USE_CERTIFICATE frame along the HTTP request. The reason I think we should change as such is, because from server's perspective, it is easier to always wait for USE_CERTIFICATE frame rather than figuring out if the frame has been omitted.